language/vm/src/deserializer.rs

```diff
@@ -661,44 +661,117 @@ fn load_signature_tokens(cursor: &mut Cursor<&[u8]>) -> BinaryLoaderResult<Vec<S
-            SerializedType::BOOL => Ok(SignatureToken::Bool),
-            SerializedType::U8 => Ok(SignatureToken::U8),
-            SerializedType::U64 => Ok(SignatureToken::U64),
-            SerializedType::U128 => Ok(SignatureToken::U128),
-            SerializedType::ADDRESS => Ok(SignatureToken::Address),
-            SerializedType::VECTOR => {
-                let ty = load_signature_token(cursor)?;
-                Ok(SignatureToken::Vector(Box::new(ty)))
-            }
-            SerializedType::REFERENCE => {
-                let ref_token = load_signature_token(cursor)?;
-                Ok(SignatureToken::Reference(Box::new(ref_token)))
-            }
-            SerializedType::MUTABLE_REFERENCE => {
-                let ref_token = load_signature_token(cursor)?;
-                Ok(SignatureToken::MutableReference(Box::new(ref_token)))
+    use SerializedType as S;
+
+    enum TypeConstructor {
+        Saturated(SignatureToken),
+        Vector,
+        Reference,
+        MutableReference,
+        StructInst {
+            sh_idx: StructHandleIndex,
+            arity: usize,
+            ty_args: Vec<SignatureToken>,
+        },
+    }
+
+    impl TypeConstructor {
+        fn apply(self, tok: SignatureToken) -> Self {
+            match self {
+                T::Vector => T::Saturated(SignatureToken::Vector(Box::new(tok))),
+                T::Reference => T::Saturated(SignatureToken::Reference(Box::new(tok))),
+                T::MutableReference => {
+                    T::Saturated(SignatureToken::MutableReference(Box::new(tok)))
+                }
+                T::StructInst {
+                    sh_idx,
+                    arity,
+                    mut ty_args,
+                } => {
+                    ty_args.push(tok);
+                    if ty_args.len() >= arity {
+                        T::Saturated(SignatureToken::StructInstantiation(sh_idx, ty_args))
+                    } else {
+                        T::StructInst {
+                            sh_idx,
+                            arity,
+                            ty_args,
+                        }
+                    }
+                }
+                _ => unreachable!("invalid type constructor application"),
             }
-            SerializedType::STRUCT => {
-                let sh_idx = read_uleb_u16_internal(cursor)?;
-                Ok(SignatureToken::Struct(StructHandleIndex(sh_idx)))
+        }
+
+        fn is_saturated(&self) -> bool {
+            match self {
+                T::Saturated(_) => true,
+                _ => false,
             }
-            SerializedType::STRUCT_INST => {
-                let sh_idx = read_uleb_u16_internal(cursor)?;
-                let type_params = load_signature_tokens(cursor)?;
-                Ok(SignatureToken::StructInstantiation(
-                    StructHandleIndex(sh_idx),
-                    type_params,
-                ))
+        }
```

Comment:
You should be able to use the new `matches!` macro for this case

Reply:
`matches!(self, T::Saturated(_))`